engine/engine/content/builtins/scripts/edn.lua

```diff
@@ -68,38 +55,34 @@ local function encode_number(val)
 end
 
 local function encode_string(val)
-  return '"' .. val:gsub('[%z\1-\31\\"]', escape_char) .. '"'
+  return '"'..val:gsub('[%z\1-\31\\"]', escape_char)..'"'
 end
 
-local function encode_as_tagged_string(tag, val)
-  return tag .. encode_string(str(val))
+local function encode_table(val)
+  return string.format('#lua/ref"%p"', val)
 end
 
-local function make_tagged_string_encoder(tag)
-  return function(val)
-    return encode_as_tagged_string(tag, val)
-  end
+function encode_userdata(val)
+  return '#lua/userdata"'..str(val)..'"'
 end
 
-local function get_table_ref(t)
-  return get_prefixed_addr("table: ", t)
+function encode_function(val)
+  return '#lua/function"'..str(val)..'"'
 end
 
-local function make_ref_encoder(get_ref)
-  return function(val)
-    return encode_as_tagged_string("#lua/ref", get_ref(val))
-  end
+function encode_thread(val)
+  return '#lua/thread"'..str(val)..'"'
 end
 
 local type_to_primitive_encoder = {
   ["nil"] = encode_nil,
   ["boolean"] = tostring,
   ["number"] = encode_number,
   ["string"] = encode_string,
-  ["table"] = make_ref_encoder(get_table_ref),
-  ["userdata"] = make_tagged_string_encoder("#lua/userdata"),
-  ["function"] = make_tagged_string_encoder("#lua/function"),
-  ["thread"] = make_tagged_string_encoder("#lua/thread"),
```

Comment:
functional-like code looks nicer, but on a big count of operations it works slower because adds extra scopes should be kept and extra function calls
+ generalisation of everything adds a few heavycalls in places where it's not necessary (see comment in `encode_function`)